torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -206,47 +206,54 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::sendImpl(
-
-        std::vector<torch::Tensor> preamble = {torch::tensor(
-            {(int64_t)pg_->getRank(),
-             (int64_t)serializedPayload.length(),
-             (int64_t)work.message_.type()},
-            {torch::kLong})};
-
-        // ProcessGroup is not thread-safe when sending with the same tag, hence
-        // the lock
-        std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
-        const auto& dst = work.to_.id_;
-        if (work.message_.isShutdown()) {
-          pendingSends.reserve(1);
-          std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-          pendingSends.emplace_back(
-              pg_->send(preamble, dst, dst /* channelTag */));
-        } else {
-          std::vector<torch::Tensor> payload = {torch::from_blob(
-              (void*)serializedPayload.c_str(),
-              serializedPayload.length(),
-              {torch::kChar})};
-          pendingSends.reserve(2);
-          std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-          pendingSends.emplace_back(
-              pg_->send(preamble, dst, dst /* channelTag */));
-          pendingSends.emplace_back(
-              pg_->send(payload, dst, dst /* channelTag */));
-        }
-        for (auto& pendingSend : pendingSends) {
-          pendingSend->wait();
+        try {
+          std::stringstream ss;
+          serialize(work.message_, ss);
+          std::string serializedPayload = ss.str();
+
+          std::vector<torch::Tensor> preamble = {torch::tensor(
+              {(int64_t)pg_->getRank(),
+               (int64_t)serializedPayload.length(),
+               (int64_t)work.message_.type()},
+              {torch::kLong})};
+
+          // ProcessGroup is not thread-safe when sending with the same tag,
+          // hence the lock
+          std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
+          const auto& dst = work.to_.id_;
+          if (work.message_.isShutdown()) {
+            pendingSends.reserve(1);
+            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+            pendingSends.emplace_back(
+                pg_->send(preamble, dst, dst /* channelTag */));
+          } else {
+            std::vector<torch::Tensor> payload = {torch::from_blob(
+                (void*)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar})};
+            pendingSends.reserve(2);
+            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+            pendingSends.emplace_back(
+                pg_->send(preamble, dst, dst /* channelTag */));
+            pendingSends.emplace_back(
+                pg_->send(payload, dst, dst /* channelTag */));
+          }
+          for (auto& pendingSend : pendingSends) {
+            pendingSend->wait();
+          }
+          future->markCompleted();
+        } catch (...) {
+          future->markCompleted(Message({}, {}, MessageType::EXCEPTION));
```

Comment:
This hides the underlying cause. Always make it visible. This mutes programming errors.

Reply:
cc @pritamdamania87 , is this what you meant by catching the exception and marking the future with the exception? If so, is there a way to make sure that the exception information isn't lost? I was thinking we could put the exception information into the message's payload for example.

Reply:
We can do something like this:

```
} catch (std::exception& e) {
  future->markCompleted(createException(work.message_, e));
} catch (...) {
  future->markCompleted(Message({"Unknown exception occurred"});
}
```

You can probably refactor `createException` from request_callback.cpp and re-use it here.